### ISSUE TITLE
Map Reduce orchestration question

### ISSUE THREAD

**Issue by Sovetnikov** [celery/celery#4293]
I'm trying to implement simple Map Reduce pattern with celery.
1. Take some initial data array = start task get_ids()
2. Start several tasks in parallel, to map each array element to some calculation result = chain results of get_ids() to split_tasks() and start chord with aggregate_result() as callback
3. Aggregate results to single result = make aggregation in aggregate_result() with chord head results

My tasks is long running (few minutes) and memory intensive in production, my celery setup has CELERYD_MAX_TASKS_PER_CHILD=1 and CELERYD_PREFETCH_MULTIPLIER=1 with -Ofair

Problems that i'm facing when trying to get final result of chord callback:
1. AsyncResult.get() completes when last task in chord head completes, not when chord callback completes
2. AsyncResult.get() returns empty result when it completes and only few seconds later (when chord callback actually completes) it returns actual result of chord callback
3. AsyncResult.get_leaf() returns result of last chord head task, not of chord callback task result

Complete test case that demonstrates my case is below.

Celery version: 3.1.25
RabbitMQ as a broker, and memcached as results backend

Main question is how to wait chord callback task completion in this situation?

```
import os
from datetime import timedelta
from subprocess import Popen
from time import sleep

from celery import Celery
from celery import chain, chord
from django.test import TestCase

ABSOLUTE_PATH = lambda x: os.path.abspath(os.path.join(os.path.abspath(os.path.dirname(__file__)), x))

app = Celery('celery_test')
app.autodiscover_tasks((__name__,))

app.conf.update(
    BROKER_URL='amqp://localhost:5672',
    CELERY_RESULT_BACKEND='cache+memcached://127.0.0.1:11211/',
    CELERY_TASK_RESULT_EXPIRES=timedelta(days=1),
    CELERY_SEND_TASK_ERROR_EMAILS=False,
    CELERYD_LOG_COLOR=False,
    CELERY_ALWAYS_EAGER=False,
    CELERY_EAGER_PROPAGATES_EXCEPTIONS=True,
    CELERYD_MAX_TASKS_PER_CHILD=1,
    CELERYD_PREFETCH_MULTIPLIER=1,
    CELERY_TASK_SERIALIZER='pickle',
    CELERY_ACCEPT_CONTENT=['pickle',],
    CELERY_RESULT_SERIALIZER='pickle',
)

if __name__ == '__main__':
    from celery.bin.celery import main as celery_main
    celery_main()


@app.task
def get_ids():
    return [3, 4, 5, 6]


@app.task
def get_id_result(id, n):
    return id * n


@app.task
def split_tasks(ids):
    st = []
    for id in ids:
        for n in range(1, 3):
            st.append(get_id_result.s(id, n))
    return chord(st)(aggregate_result.s())


@app.task
def aggregate_result(results):
    return sum(results)


@app.task
def calculate():
    return chain(get_ids.s(), split_tasks.s())()


class TestCelery(TestCase):
    def setUp(self):
        pcwd = os.getcwd()
        os.chdir(ABSOLUTE_PATH('../..'))
        cmd = 'python {1} --hostname=test-channel@host -A {0} worker -l info --concurrency=4 -Ofair'.format(__name__, __file__)
        self.celery_process = Popen(cmd)
        os.chdir(pcwd)

    def tearDown(self):
        self.celery_process.kill()

    def test_celery(self):
        r = calculate.delay() # this line returns right AsyncResult with task id of chord callback
        get_result = r.get() # get() returns too early, before aggregate_result() is executed
        collect_result = list(r.collect())
        leaf_result = r.get_leaf()

        result = r.get().result.result # result is None at moment when get() returns
        self.assertEqual(result, 54) # result is 54 as expected if i wait manually when aggregate_result() completes

        while True:
            sleep(2)
```

**Comment by Sovetnikov** [comment]
Just get the point myself.
Because my tasks returns AsyncResult of subtasks i need to wait all AsyncResults:
calculate.delay().get().get().get()

But i still does not understand why get_leaf() returns wrong result.

Another question about mapping some task results to another task.
Celery canvas offers Task.map(collection) and xmap(Signature, collection) that take collection and task signature that will be started for every element of collection.
But i have only task that will return collection for mapping instead of collection itself.
How i can connect these two tasks with "map"?

**Comment by georgepsarakis** [comment]
The [chord](http://docs.celeryproject.org/en/latest/userguide/canvas.html#chords) can possibly yield a pattern similar to MapReduce. See the example in the documentation.

**Comment by Sovetnikov** [comment]
I read celery canvas guide many times and you can see from my example that i know how to use chord, thanks.

**Comment by Sovetnikov** [comment]
AsyncResult.get_leaf() works unexpectedly because of https://github.com/celery/celery/issues/703

**Comment by Sovetnikov** [comment]
As an end note.

Celery 3 does not support full stack declarative distributed task orchestration (main word "declarative"). It has join implemented as chord, but lacks fork ...

I have some success to implement fork verb for celery canvas to make possible this declarative construction:
```
get_initial_data.s() | dmap(map_data_item.s()) | aggregate_result.s()
```
But some historical (strange) code breaks all in case group is an input to dmap (See commit with small commetns from @ask https://github.com/celery/celery/commit/c3c32bb7265c4302d86edb72899ae07040ff98ee):
```
# This does not work
group(get_initial_data.s(), get_initial_data.s()) | dmap(map_data_item.s()) | aggregate_result.s()
# Equivalent of (auto converted by celery)
chord(group(get_initial_data.s(), get_initial_data.s()), body=dmap(map_data_item.s()) | aggregate_result.s()
```
(my dmap implementation is at the end of this comment)

At now i must stick to Celery 3 because a lots of dependencies that must be upgraded in my project with migration to Celery 4 and that migrations does not worth the benefits.
As i see Celery 4 still has no fork verb in canvas, it is strange to me. One thing that makes life easier with Celery 4 is Task.replace that makes possible to substitute one task results with another and make some helpers or canvas verbs.

My case divides on two cases:
1. Make tasks run in right sequence.
This is hard to implement declaratively with canvas verbs, but this possible with tasks that triggers another tasks.
2. Get result of final task
You can return AsyncResult of task that current task will trigger next, and you can get to the final AsyncResult that will return actual value.
Or, the good thing is that with Celery you can get final task id even before any tasks in your workflow is submitted to queue.

Some solutions with Celery 3:
1. As described in my test case above, just need to make .get().get().get(). One can try to implement some customized AsyncResult that will hide this logic underneath API.
2. With some helpers and possible to get final task id in advance here is solution that looks like "declarative":
```
@app.task
def get_data(*args, **kwargs):
    return [3, 4, 5, 6]


@app.task(bind=True)
def map_data_item(self, n, id):
    return id * n


@app.task(bind=True)
def aggregate_result(self, results):
    return sum(flatten(results))


@app.task(bind=True)
def dmap_task(self, it, mapper_task, reduce_task):
    mapper = maybe_signature(mapper_task, self.app)
    reduce = maybe_signature(reduce_task, self.app)
    st = []
    for item in it:
        st.append(mapper.clone(args=(item,)))
    return chord(st)(reduce)


def calculate():
    # Final task signature
    final = aggregate_result.s()
    # Here we got final task AsyncResult with id that will be assigned to task when it will be executed
    final_result = final.freeze()
    # Starting our workflow with final task at end
    chain(get_data.s(), dmap_task.s(map_data_item.s(3), final))()
    # Just return final task AsyncResult
    return final_result

if __name__ == '__main__':
    final_result_value = calculate().get()
```

My dmap implementation based on group, if someone interested:
```
@Signature.register_type
class dmap(Signature):
    def __init__(self, mapper_task, **options):
        self.mapper_task = mapper_task
        Signature.__init__(
            self, 'celery.dmap', (), {'mapper_task': mapper_task}, **options
        )

    def apply_async(self, args=(), kwargs=None, add_to_parent=True, **options):
        tasks = _maybe_clone(self.tasks, app=self._app)
        if not tasks:
            return self.freeze()
        type = self.type
        return type(*type.prepare(dict(self.options, **options), tasks, args),
                    add_to_parent=add_to_parent)

    def apply(self, *args, **kwargs):
        if not self.tasks:
            return self.freeze()  # empty group returns GroupResult
        return Signature.apply(self, *args, **kwargs)

    def __call__(self, *partial_args, **options):
        return self.apply_async(partial_args, **options)

    def freeze(self, _id=None, group_id=None, chord=None):
        opts = self.options
        try:
            gid = opts['task_id']
        except KeyError:
            gid = opts['task_id'] = uuid()
        if group_id:
            opts['group_id'] = group_id
        if chord:
            opts['chord'] = group_id
        new_task = maybe_signature(self.mapper_task, app=self._app).clone()

        self.mapper_task = self.kwargs['mapper_task'] = new_task
        return self.app.GroupResult(gid, [])

    _freeze = freeze

    @property
    def app(self):
        return self._app or (self.mapper_task.app if self.mapper_task else current_app)

    @property
    def type(self):
        if self._type:
            return self._type
        # taking the app from the first task in the list, there may be a
        # better solution for this, e.g. to consolidate tasks with the same
        # app and apply them in batches.
        return self.app.tasks[self['mapper_task']]


@connect_on_app_finalize
def add_dmap_task(app):
    from celery.canvas import maybe_signature
    from celery import chord as canvas_chord

    _app = app

    class DMap(app.Task):
        app = _app
        name = 'celery.dmap'
        accept_magic_kwargs = False
        _decorated = True

        def apply_async(self, args=(), kwargs={}, group_id=None, chord=None,
                        task_id=None, link=None, link_error=None, **options):
            if len(link) > 1:
                raise Exception('dmap supports only one link')
            mapper = maybe_signature(kwargs['mapper_task'], self.app)
            it = flatten(args)
            tasks = []
            for item in it:
                tasks.append(mapper.clone(args=(item,)))
            callback = maybe_signature(link[0], self.app)
            c = canvas_chord(tasks, body=callback)
            if isinstance(c, AsyncResult):
                return c
            return c()

    return DMap
```

**Comment by thedrow** [comment]
I haven't read the issue but just popping here to note that Celery is not really meant for map-reduce operations. It's a task manager.
Use another component for these kind of workloads.

@georgepsarakis If there's no actionable item, please close this issue.

**Comment by auvipy** [comment]
better to post discussion type issues to mailing list.
